### ISSUE TITLE
Different results from using scale_pos_weight vs oversampling

### ISSUE THREAD

**Issue by mchav** [dmlc/xgboost#11516]
I get different results when I take a toy dataset and set `scale_pos_weight = 10`  versus taking that same dataset and oversampling the positive class by a factor of 10. I assumed that these two approaches would yield the same results. I can't seem to find any documentation about the difference between `scale_pos_weight` and oversampling. Is there anything I'm missing?

NB: I get a similar difference when I run SparkXGBClassifier with a weight column that's 10x for the positive class versus when I run with `scale_pos_weight`.

I have a synthetic, toy dataset generated as follows:

```
training_rows = []


for i in range(1000000):
    row = {
        ## A correlates with our label.
        "A": (-1 if i % 2 == 0 else 1) * random.random(),
        ## B is an integer similar to A - I'd like to see if xgboost thinks it's more important.
        "B": round((-100 if i % 2 == 1 else 100) * random.random()),
        ## C is a constant and shouldn't be important
        "C": 5,
        ## D is entirely random and shouldn't be important.
        "D": random.random(),
        "E": True if random.random() > 0.8 else False,
        "F": 1 if i % 2 == 0 else 0}
    training_rows.append(row)
```

And the model is trained as follows:

```
xgb_params = {'features_col': 'features',
              'label_col': 'F',
              'validation_indicator_col': 'E',
              'weight_col': 'weight',
              'eval_metric': 'aucpr',
              'early_stopping_rounds': 10,
              'num_workers': sc.defaultParallelism,
              'tree_method': 'hist',
              'n_estimators': 100,
              'seed': 123,
              }


classifier = SparkXGBClassifier(**xgb_params)

xgb_model = classifier.fit(assembled_train_val)

print(xgb_model.get_booster().best_iteration)
print(xgb_model.get_booster().best_score)
xgb_model.get_booster().feature_names = features
x = xgb_model.get_feature_importances(importance_type='gain')

{k: v for k, v in sorted(x.items(), key=lambda item: item[1], reverse=True)}
```

For the scale_pos_weight I replace 'weight_col' with scale_pos_weight and for oversampling I remove both these parameters.

**Comment by chyunsu3** [comment]
`scale_pos_weight` is not implemented with oversampling. It is implemented by assigning more weights to positive samples when computing the loss.

**Comment by mchav** [comment]
Shouldn't that make the implementations somewhat similar? And does that mean that pyspark's weight column is implemented with oversampling?

**Comment by chyunsu3** [comment]
> pyspark's weight column is implemented with oversampling?

No, the weight column works the same way as `scale_pos_weight`. In fact, `scale_pos_weight` works by scaling the weights of the positive samples by a fixed amount:
https://github.com/dmlc/xgboost/blob/68182c26503083a17bf06ecc530cf09c5d0dd942/src/objective/regression_obj.cu#L157-L163

**Comment by mchav** [comment]
For context, we are trying to refactor from using scale_pos_weight to using instance weights and the results from each approach look different. and I'm trying to understand why.

From another small experiment these two methods (using weights directly vs using instance weights) don't give the same tree even when controlling for randomness with a seed. Some assumptions I'm making:
* the runs are mostly deterministic meaning there should be small differences between runs
* if two configurations are equivalent they should produce similar importance gains on features.

Are these the assumptions wrong? If not, why do we get different looking models?

Given the following code:

```python
import pandas as pd
import numpy as np
import xgboost as xgb

import random

training_rows = []


for i in range(1000000):
    row = {
        ## A correlates with our label.
        "A": (-1 if i % 2 == 0 else 1) * random.random(),
        ## B is an integer similar to A - I'd like to see if xgboost thinks it's more important.
        "B": round((-100 if i % 2 == 1 else 100) * random.random()),
        ## C is a constant and shouldn't be important
        "C": 5,
        ## D is entirely random and shouldn't be important.
        "D": random.random(),
        ## Validation
        "E": True if random.random() > 0.8 else False,
        ## Weight
        "W": 10 if i % 2 == 0 else 1,
        "F": 1 if i % 2 == 0 else 0}
    training_rows.append(row)

training_df = pd.DataFrame(training_rows)

features = ["A", "B", "C", "D"]

X_train = training_df[training_df['E'] == False][features]
y_train = training_df[training_df['E'] == False]["F"]
w_train = training_df[training_df['E'] == False]["W"]

X_test = training_df[training_df['E'] == True][features]
y_test = training_df[training_df['E'] == True]["F"]
w_test = training_df[training_df['E'] == True]["W"]

dtrain = xgb.DMatrix(X_train, label=y_train,  feature_names=X_train.columns.to_list(), weight=w_train)

dtest = xgb.DMatrix(X_test, label=y_test, feature_names=X_test.columns.to_list(), weight=w_test)


xgb_params = {'eval_metric': 'aucpr',
              'tree_method': 'hist',
              'seed': 123,
              }

eval_set = [(dtrain, "train"), (dtest, "validation")]

num_round = 100
bst = xgb.train(xgb_params, dtrain, num_boost_round=num_round, evals=eval_set, early_stopping_rounds=10)

print(bst.best_iteration)
print(bst.best_score)
bst.feature_names = features
x = bst.get_score(importance_type='gain')

print({k: v for k, v in sorted(x.items(), key=lambda item: item[1], reverse=True)})

## ------------- Now with no weights ---------------------- ##
dtrain = xgb.DMatrix(X_train, label=y_train,  feature_names=X_train.columns.to_list())

dtest = xgb.DMatrix(X_test, label=y_test, feature_names=X_test.columns.to_list())


xgb_params = {'eval_metric': 'aucpr',
              'tree_method': 'hist',
              'scale_pos_weight': 10,
              'seed': 123,
              }

eval_set = [(dtrain, "train"), (dtest, "validation")]

num_round = 100
bst = xgb.train(xgb_params, dtrain, num_boost_round=num_round, evals=eval_set, early_stopping_rounds=10)

print(bst.best_iteration)
print(bst.best_score)
bst.feature_names = features
x = bst.get_score(importance_type='gain')

print({k: v for k, v in sorted(x.items(), key=lambda item: item[1], reverse=True)})
```

I get the following results:

```
(ml_testbed) $ % python3 weight_test.py
[0]	train-aucpr:1.00000	validation-aucpr:1.00000
[1]	train-aucpr:1.00000	validation-aucpr:1.00000
[2]	train-aucpr:1.00000	validation-aucpr:1.00000
[3]	train-aucpr:1.00000	validation-aucpr:1.00000
[4]	train-aucpr:1.00000	validation-aucpr:1.00000
[5]	train-aucpr:1.00000	validation-aucpr:1.00000
[6]	train-aucpr:1.00000	validation-aucpr:1.00000
[7]	train-aucpr:1.00000	validation-aucpr:1.00000
[8]	train-aucpr:1.00000	validation-aucpr:1.00000
[9]	train-aucpr:1.00000	validation-aucpr:1.00000
0
0.9999999999849465
{'B': 32235.7578125, 'A': 175.5067901611328, 'D': 0.9925268292427063}
[0]	train-aucpr:1.00000	validation-aucpr:1.00000
[1]	train-aucpr:1.00000	validation-aucpr:1.00000
[2]	train-aucpr:1.00000	validation-aucpr:1.00000
[3]	train-aucpr:1.00000	validation-aucpr:1.00000
[4]	train-aucpr:1.00000	validation-aucpr:1.00000
[5]	train-aucpr:1.00000	validation-aucpr:1.00000
[6]	train-aucpr:1.00000	validation-aucpr:1.00000
[7]	train-aucpr:1.00000	validation-aucpr:1.00000
[8]	train-aucpr:1.00000	validation-aucpr:1.00000
[9]	train-aucpr:1.00000	validation-aucpr:1.00000
0
1.0
{'A': 64797.8984375, 'B': 27.068464279174805}
```

**Comment by chyunsu3** [comment]
@trivialfis Does the quantile sketch process honor `scale_pos_weight`? In the following snippet, the quantile sketcher uses `info.weights_` but not `scale_pos_weight`:
https://github.com/dmlc/xgboost/blob/68182c26503083a17bf06ecc530cf09c5d0dd942/src/common/quantile.cc#L64-L87

@mchav I suspect that the difference comes from the way XGBoost computes approximate quantiles of data. The quantile sketch is using instance weights but (probably) not `scale_pos_weight`.

**Comment by trivialfis** [comment]
No, the `scale_pos_weight` is very specific to the logistic objective.

**Comment by mchav** [comment]
@trivialfis 

I see. So we don't expect scale_pos_weight in the binary logistics case to be equivalent to setting instance weights?

And scale_pos_weight only affects the loss function as opposed to everything else about the model (cost, regularization params etc?). If this is the case is there a document that describes how these would differ. I think it's reasonable to think they would be functionally equivalent.

**Comment by trivialfis** [comment]
Indeed. That parameter has been there since forever, we should either create alternative like class weight or document it more precisely.

**Comment by mchav** [comment]
Alright. Thank you for the quick turn around on this.

This explains why experiments from using instance weights produced results that looked like what we got from oversampling. A team mate gave the following explanation that we wanted to confirm:

scale_pos_weight: applied globally during gradient and Hessian calculations. Affects the loss function uniformly for all positive samples. Implemented as a scaling factor in the objective function. Preserves the original sample distribution while reweighing the loss. 

Instance weights: applied individually to each sample during training, changes the effective sample distribution that the algorithm sees. Affects tree splitting criteria differently. Can create different regularization effects. 

Is this a fair summary of the differences?

If so, it might be good to put a side by side comparison like this in the documentation. 

RE class weights: that would be great and I think captures the intuition that I originally associated with scale_pos_weight.

**Comment by trivialfis** [comment]
> Is this a fair summary of the differences?

Looks good.

> RE class weights: that would be great and I think captures the intuition that I originally associated with scale_pos_weight.

Yeah, we should focus on this instead of creating more confusion by introducing difficult to explain documents.

**Comment by mchav** [comment]
I'm not sure what the API for this would look like and how it would interact with the weights passed into the dmatrix.

The weights are currently passed to MetaInfo as an instance variable then used in various places after that. My initial thought was to add a getter method for the weights that returns the user defined weights if they exist or, if they don't exist, create a new collection of weights based on the labels and a new class_weights instance variable. Then refactor all instances of info.weights_ to info.GetWeights().

This approach requires the user to know that one option overrides the other. Unless the two fields combine multiplicatively like scale weight and instance weights. What do you think @trivialfis ?

**Comment by trivialfis** [comment]
I can't be sure yet. To me, the easiest way to do it is simply remove the parameter from XGBoost and rely on user feature engineering process to create sample weights based on sample class.

**Comment by mchav** [comment]
Not sure about that. There’s still utility in specifying weights for a  entire class without having to recompute a column in a dataframe. In fact, when experimenting with weights one usually tries out different weights on the same dataset.

**Comment by mchav** [comment]
@trivialfis it'd be great to close on this. Is the recoemmendation still to deprecate scale_pos_weight?

I'm updating this document and noticed that the recommendation here might be misguided: https://fraud-detection-handbook.github.io/fraud-detection-handbook/Chapter_6_ImbalancedLearning/Ensembling.html#weighted-xgboost
